theme-kit/router.ts

```diff
@@ -0,0 +1,147 @@
+/* eslint-disable @typescript-eslint/no-use-before-define,@typescript-eslint/naming-convention */
+
+import type { Culture } from '@prezly/sdk';
+import { notFound } from 'next/navigation';
+import type { ReactElement, ReactNode } from 'react';
+import UrlPattern from 'url-pattern';
+
+import { assertServerEnv } from '@/utils';
+
+import { api } from './api';
+
+export function route<
+    Pattern extends `/${string}`,
+    Match extends ExtractPathParams<Pattern>,
+    Props = Match,
+>(pattern: Pattern, loader: Loader<Match, Props>) {
+    const urlPattern = buildUrlPattern(pattern);
+
+    return {
+        async match(path: string, getDefaultLocale, isEnabledLocale) {
+            const matched = urlPattern.match(path);
+
+            if (!matched) {
+                return undefined;
+            }
+
+            if (matched.locale && !(await isEnabledLocale(matched.locale))) {
+                return undefined;
+            }
+
+            if (!matched.locale) {
+                matched.locale = await getDefaultLocale();
+            }
```

Comment:
Locale code matching and defaults is baked into the router, so it doesn't have to be implemented for every localized sub-path.